clients/src/main/java/org/apache/kafka/clients/consumer/internals/StreamsMembershipManager.java

```diff
@@ -0,0 +1,1284 @@
+        return currentState != MemberState.RECONCILING && nextState == MemberState.RECONCILING;
+    }
+
+    private void resetEpoch() {
+        updateMemberEpoch(StreamsGroupHeartbeatRequest.JOIN_GROUP_MEMBER_EPOCH);
+    }
+
+    private void updateMemberEpoch(int newEpoch) {
+        boolean newEpochReceived = this.memberEpoch != newEpoch;
+        this.memberEpoch = newEpoch;
+        if (newEpochReceived) {
+            if (memberEpoch > 0) {
+                notifyEpochChange(Optional.of(memberEpoch));
+            } else {
+                notifyEpochChange(Optional.empty());
+            }
+        }
+    }
+
+    /**
+     * Discard assignments received that have not been reconciled yet (waiting for metadata
+     * or the next reconciliation loop).
+     */
+    private void clearCurrentTaskAssignment() {
+        currentAssignment = LocalAssignment.NONE;
+    }
+
+    /**
+     * Clear the assigned partitions in the member subscription, pending assignments and metadata cache.
+     */
+    private void clearTaskAndPartitionAssignment() {
+        subscriptionState.assignFromSubscribed(Collections.emptySet());
+        notifyAssignmentChange(Collections.emptySet());
+        currentAssignment = LocalAssignment.NONE;
+        targetAssignment = LocalAssignment.NONE;
+    }
+
+    /**
+     * @return True if the member should not send heartbeats, which is the case when it is in a
+     * state where it is not an active member of the group.
+     */
+    public boolean shouldSkipHeartbeat() {
+        return isNotInGroup();
+    }
+
+    /**
+     * @return True if the member should send heartbeat to the coordinator without waiting for
+     * the interval.
+     */
+    public boolean shouldHeartbeatNow() {
+        return state == MemberState.ACKNOWLEDGING || state == MemberState.LEAVING || state == MemberState.JOINING;
+    }
+
+    /**
+     * Set {@link #subscriptionUpdated} to true to indicate that the subscription has been updated.
+     * The next {@link #onConsumerPoll()} will join the group with the updated subscription, if the member is not part of it yet.
+     * If the member is already part of the group, this will only ensure that the updated subscription
+     * is included in the next heartbeat request.
+     * <p/>
+     * Note that the list of topics in the subscription is taken from the shared subscription state.
+     */
+    public void onSubscriptionUpdated() {
+        subscriptionUpdated.compareAndSet(false, true);
+    }
+
+    /**
+     * Join the group if the member is not part of it yet. This function separates {@link #transitionToJoining}
+     * from the {@link #onSubscriptionUpdated} to fulfill the requirement of the "rebalances will only occur during an
+     * active call to {@link org.apache.kafka.clients.consumer.KafkaConsumer#poll(java.time.Duration)}"
+     */
+    public void onConsumerPoll() {
+        if (subscriptionUpdated.compareAndSet(true, false) && state == MemberState.UNSUBSCRIBED) {
```

Comment:
In the test this part always evaluates the `true` case, never false, probably would be good to add a case for the `false` case

Reply:
I added some tests.